File: actin-common/src/test/java/com/hartwig/actin/clinical/feed/questionnaire/QuestionnaireVersionTest.java
package com.hartwig.actin.clinical.feed.questionnaire;

import static org.junit.Assert.assertEquals;

import org.jetbrains.annotations.NotNull;
import org.junit.Test;

public class QuestionnaireVersionTest {

    @Test
    public void canResolveAllVersions() {
        assertEquals(QuestionnaireVersion.V1_0B,
                QuestionnaireVersion.version(entry(TestQuestionnaireFactory.createTestQuestionnaireValueV1_0B())));
        assertEquals(QuestionnaireVersion.V1_0A,
                QuestionnaireVersion.version(entry(TestQuestionnaireFactory.createTestQuestionnaireValueV1_0A())));
        assertEquals(QuestionnaireVersion.V0,
                QuestionnaireVersion.version(entry(TestQuestionnaireFactory.createTestQuestionnaireValueV0())));
    }

    @NotNull
    private static QuestionnaireEntry entry(@NotNull String questionnaire) {
        return ImmutableQuestionnaireEntry.builder()
                .from(TestQuestionnaireFactory.createTestQuestionnaireEntry())
                .itemAnswerValueValueString(questionnaire)
                .build();
    }
}

File: clinical/src/test/kotlin/com/hartwig/actin/clinical/feed/emc/questionnaire/QuestionnaireVersionTest.kt
package com.hartwig.actin.clinical.feed.emc.questionnaire

import com.hartwig.actin.clinical.feed.emc.questionnaire.QuestionnaireVersion.Companion.version
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class QuestionnaireVersionTest {

    @Test
    fun `Should resolve all versions`() {
        assertVersion(QuestionnaireVersion.V1_7, TestQuestionnaireFactory.createTestQuestionnaireValueV1_7())
        assertVersion(QuestionnaireVersion.V1_6, TestQuestionnaireFactory.createTestQuestionnaireValueV1_6())
        assertVersion(QuestionnaireVersion.V1_5, TestQuestionnaireFactory.createTestQuestionnaireValueV1_5())
        assertVersion(QuestionnaireVersion.V1_4, TestQuestionnaireFactory.createTestQuestionnaireValueV1_4())
        assertVersion(QuestionnaireVersion.V1_3, TestQuestionnaireFactory.createTestQuestionnaireValueV1_3())
        assertVersion(QuestionnaireVersion.V1_2, TestQuestionnaireFactory.createTestQuestionnaireValueV1_2())
        assertVersion(QuestionnaireVersion.V1_1, TestQuestionnaireFactory.createTestQuestionnaireValueV1_1())
        assertVersion(QuestionnaireVersion.V1_0, TestQuestionnaireFactory.createTestQuestionnaireValueV1_0())
        assertVersion(QuestionnaireVersion.V0_2, TestQuestionnaireFactory.createTestQuestionnaireValueV0_2())
        assertVersion(QuestionnaireVersion.V0_1, TestQuestionnaireFactory.createTestQuestionnaireValueV0_1())
    }

    @Test(expected = IllegalStateException::class)
    fun `Should throw exception on unresolved version`() {
        version(TestQuestionnaireFactory.entryWithText("Not an entry"))
    }

    private fun assertVersion(expected: QuestionnaireVersion, questionnaire: String) {
        assertThat(version(TestQuestionnaireFactory.entryWithText(questionnaire))).isEqualTo(expected)
    }
}